packages/data/src/components/use-select/index.js

```diff
@@ -0,0 +1,101 @@
+/**
+ * WordPress dependencies
+ */
+import { createQueue } from '@wordpress/priority-queue';
+import {
+	useLayoutEffect,
+	useEffect,
+	useState,
+	useRef,
+	useMemo,
+	useCallback,
+} from '@wordpress/element';
+import { isShallowEqualObjects } from '@wordpress/is-shallow-equal';
+
+/**
+ * Internal dependencies
+ */
+import useRegistry from '../registry-provider/use-registry';
+import useAsyncMode from '../async-mode-provider/use-async-mode';
+
+/**
+ * Favor useLayoutEffect to ensure the store subscription callback always has
+ * the selector from the latest render. If a store update happens between render
+ * and the effect, this could cause missed/stale updates or inconsistent state.
+ *
+ * Fallback to useEffect for server rendered components because currently React
+ * throws a warning when using useLayoutEffect in that environment.
+ */
+// const useIsomorphicLayoutEffect = typeof window !== 'undefined' ?
+// 	useLayoutEffect : useEffect;
+const useIsomorphicLayoutEffect = typeof window !== 'undefined' ?
+	useLayoutEffect : useEffect;
+
+const renderQueue = createQueue();
+
+export default function useSelect( mapSelect, nextProps ) {
+	const registry = useRegistry();
+	const isAsync = useAsyncMode();
+	const [ mapOutput, setMapOutput ] = useState( mapSelect( registry.select ) );
+	const queueContext = useMemo( () => ( { queue: true } ), [ registry ] );
+
+	const latestMapSelect = useRef( mapSelect );
+	const latestIsAsync = useRef( isAsync );
+	const latestMapOutput = useRef( mapOutput );
+	const isMounted = useRef();
+
+	/**
+	 * Set current values.
+	 */
+	useIsomorphicLayoutEffect( () => {
+		latestMapSelect.current = mapSelect;
+		isMounted.current = true;
+	} );
+
+	/**
+	 * Flush queue if isAsync changes
+	 */
+	useIsomorphicLayoutEffect( () => {
+		latestIsAsync.current = isAsync;
+		renderQueue.flush( queueContext );
+	}, [ isAsync ] );
+
+	/**
+	 * Callback for doing changes to store.
+	 */
+	const onStoreChange = useCallback( () => {
+		if ( isMounted.current ) {
+			const newMapOutput = latestMapSelect.current( registry.select, registry );
+			if ( ! isShallowEqualObjects( latestMapOutput.current, newMapOutput ) ) {
+				latestMapOutput.current = newMapOutput;
+				setMapOutput( newMapOutput );
+			}
+		}
+	}, [ registry, nextProps ] );
+
+	/**
+	 * Map output should always be called if incoming prop dependencies
+	 * has changed.
+	 */
+	useIsomorphicLayoutEffect( () => {
+		onStoreChange();
+	}, [ registry, nextProps ] );
```

Comment:
`onStoreChange` can be the dependency here.

Reply:
It seems that this means this hooks makes the "nextProps" (the dependencies of the callback) mandatory. I'd think that we should also support a naive version where the mapping is attempted on each render if `nextProps` is undefined.

Reply:
>It seems that this means this hooks makes the "nextProps" (the dependencies of the callback) mandatory.

`undefined === undefined`, so it will only run when the registry changes.

Reply:
I almost think it _should_ be mandatory.  This allows `useSelect` to effectively "know" the dependencies.  I'm thinking then it's possible for implementations of just the hook to describe what dependencies the provided `mapSelect` callback has.  

Thus the new `withSelect` hoc is provided all the props coming in from the parent as dependencies (simply because there's no way to know what mappers will be provided by the components composed with `withSelect`).

Reply:
> I almost think it should be mandatory. 

I agree with that but conceptually speaking, it behaves differently than the `useEffect` dependencies and could mislead people.

`useEffect( something )` updates on each render, I expect
`useSelect( something )` to update on each render as well.

Reply:
`nextProps` is problematic, because it's an object so you put the burden on the user to memoize it.

`onStoreChange` doesn't need to be memoized. It can just read refs.

`nextProps` should be an array that gets passed as the dependency array to the `useIsomorphicLayoutEffect` of this chat thread.

Reply:
You also don't want to run this during the first render, so this should come before the effect that sets `isMounted`, and it should check for it to be true.

Reply:
There was another comment attached to an outdated code snippet about assuming `nextProps` is an array and spreading it here (thus being more consistent with dependency shape with `useEffect`).  I think that's probably what we should do here and have `withSelect` do the necessary implementation to follow that signature expectation for `useSelect`.

Reply:
Yeah that's better, but it still means the component re-renders twice. First when nextProps change and then after the selector is run.

Take a look at this approach: https://github.com/reduxjs/react-redux/blob/v7-hooks-alpha/src/hooks/useSelector.js

Here:
https://github.com/reduxjs/react-redux/blob/0e41eaeebf5d8d123daaa50a91ee1c219c4830de/src/hooks/useSelector.js#L61-L64
You would also check if `nextProps` have changed.

Reply:
This actually removed the need for `nextProps`.

I updated my gist to implement all the behavior you want. Look at how we can implement `withSelect` now.

https://gist.github.com/epiqueras/7eae39ba6b903286cf17a4907902a630